### DISCUSSION TITLE
Pinguino I2C issues

### DISCUSSION THREAD

**Discussion by ITstreet1** [chipKIT32/chipKIT-core#471] [Q&A] [answered]
I have Pinguino Micro.
While selecting this board in Olimex boards in Arduino IDE, I just can not compile a simple Adafruit ssd1306 sketch. What lib shall I use for PIC32 boards and SSD1306?

Another question. There are some examples of I2C LCD usage with chipKIT. It uses I2C_20x4 lib. Where to find this one? I searched but had no luck.

**Comment by JacobChrist** [comment]
I have used some LCD libraries with chipKIT boards.  When I have done so I usually got to the code from the supplier selling the LCDs.

**Comment by majenkotech** [comment, marked as answer]
[Cariad](https://github.com/CariadDisplayLibrary) (was DisplayCore) is the display library system for chipKIT. It includes an [SSD1306](https://github.com/CariadDisplayLibrary/SSD1306) driver that supports (amongst other things) [DTWI](https://github.com/CariadDisplayLibrary/SSD1306_DTWI).

Note: you need all three libraries: [Cariad.h](https://github.com/CariadDisplayLibrary/Cariad), [SSD1306.h](https://github.com/CariadDisplayLibrary/SSD1306) and [SSD1306_DTWI.h](https://github.com/CariadDisplayLibrary/SSD1306_DTWI) as each forms a layer in the strata that is Cariad.

**Reply by ITstreet1** [reply]
Invalid library found in C:\Users\user\Documents\Arduino\libraries\Cariad-master: Missing 'paragraph' from library
.
.
for now

**Reply by ITstreet1** [reply]
Ok. It works. Sort of...

I can get data on OLED128x64 but OLED_DEMO is not showing the rectangle properly. 
Where can I find keywords? They look like something from Adafruit GFX.

**Reply by majenkotech** [reply]
It did start out life as the Adafruit GFX but it got completely re-written and rebuilt from the ground up. It supports all that the Adafruit GFX library does, plus much much much more. It's a very complex library. I used to have part of my website dedicated to it, but that went the way of the dodo when I changed my website system to an ecommerce solution. Your best bet is to examine Cariad.h for the Cariad class - that contains all the drawing functions.

**Reply by ITstreet1** [reply]
I did check all the files. There are many interesting functions. But there is no setTextSize. ;)
It would be nice to have some sort of documentation. At least here on GitHub.

No matter, your example where it should print rect around the display doesn't do it right. The right side line is printed some few pixels from the top and the bottom. The rest of the line, some 80% of the middle, is moved all the way to the left. There are issues with triangles, and circles as well.

**Reply by majenkotech** [reply]
There is no "setTextSize" because there is fonts - use a bigger font.

The offset rectangle is probably the fault of the display (ish). There are many SSD1306 based displays, and none of them actually work at the display resolution of the SSD1306 chip - they are all some smaller subset of it, and different manufacturers have that "window" in different positions within the chip's memory. The SSD1306 driver has a function to combat that: setDisplayOffset().

```c++
tft.setDisplayOffset(3, 6);
```

Play with the numbers until you get the desired result.

**Reply by ITstreet1** [reply]
Can those numbers go in minus?

**Reply by majenkotech** [reply]
No, they can't. Do you need them to? Can you post a photo of how the display looks?

**Reply by ITstreet1** [reply]
Just asked. Not sure I need it.
